clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -604,6 +597,76 @@ protected void checkLongRunningTaskState() throws InterruptedException {
     }
   }
 
+  private boolean canSwitchToLeaderTopic(PartitionConsumptionState pcs) {
+    /**
+     * Potential risk: it's possible that Kafka consumer would starve one of the partitions for a long
+     * time even though there are new messages in it, so it's possible that the old leader is still producing
+     * after waiting for 5 minutes; if it does happen, followers will detect upstream offset rewind by
+     * a different producer host name.
+     */
+    long lastTimestamp = getLastConsumedMessageTimestamp(pcs.getPartition());
+    if (LatencyUtils.getElapsedTimeInMs(lastTimestamp) <= newLeaderInactiveTime) {
+      return false;
+    }
+
+    /**
+     * For user meta stores, it requires all messages in local VT topic to be consumed before switching to leader topic,
+     * otherwise, it can get into an error replica issue when the followings happen:
+     *
+     * 1. Consumed RT data is still less than threshold thus has not been persisted to disk {@link StoreIngestionTask#shouldSyncOffset}.
+     *    After host restarts, the RT offset retrieved from disk is stale.
+     * 2. RT is truncated due to retention and has no data message in it.
+     * 3. Meta system store switches to RT so quickly that has not yet fully consumed data from local VT.
+     *
+     * When all above conditions happen together, meta system stores switch to RT, but subscribe to a stale offset and
+     * cannot consume anything from it (empty) and update, thus
+     *    RT end offset - offset in leader replica > threshold
+     * thus replica cannot become online {@link StoreIngestionTask#isReadyToServe}.
+     */
+    if (VeniceSystemStoreUtils.isMetaStore(storeName) && !isKafkaTopicPartitionFullyConsumed(pcs)) {
```

Comment:
The other user system store (davinci push status system store) also has this problem because its promotion to leader delay is also 1s. We need to let davinci push status system store wait for VT consumption too. Could you please include davinci push system store into the fix? Basically use `isUserSystemStore` here and update the java doc & commit message.
```
if (isUserSystemStore()) {
      newLeaderInactiveTime = serverConfig.getServerSystemStorePromotionToLeaderReplicaDelayMs();
}
```

Reply:
I think what you suggested makes sense, but I feel that it would be less risky if we do it in another change mainly because that current analysis is based on the meta store, and this change might still have some risk (as it changes the conditions to switch to RT topic), so it is better to restrain its impacts to only meta store first and then expand to push status store later, if that makes sense. The expansion part should be easy as you mentioned here.

Reply:
That sounds a bit weird to me...

If we know that the issue affects stires with the 1s override on the transition delay, then we should apply the fix to all affected cases, as Shuhui suggested.

On the other hand, if we are unsure about the validity of the fix, then the ideal mitigation for this risk is not to fix only half the things that need it, but rather to put the new behavior behind a config flag, and/or to roll out gradually to a canary node and roll back in case of issues. Having affected only half the system stores will not really buy us much if we end up rolling back regardless...

Would suggest reconsidering.

Reply:
Thanks for the review, Felix. Error leader replicas in meta_store outnumbers that in push_status_store in the ratio of 24 to 2 in EI and that is main reason that I was trying to target meta store, to fix the majority first, but as you explained, it is more preferred to fix all cases, I think I will analyze push_status_store issues and verify the effectiveness with a Frankenwar in EI. (I already did the Frankenwar verification for meta_store).

Reply:
Updated the rb to make it applicable to both meta-store and davinci_push_status store. Verified the effectiveness of the fix in EI through a Frankenwar test in one of those hosts who is having error leader issues for its push_status stores.